src/MongoDB/CursorInterface.c

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <php.h>
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+
+zend_class_entry* php_phongo_cursor_interface_ce;
+
+/* {{{ MongoDB\BSON\CursorInterface function entries */
+ZEND_BEGIN_ARG_INFO_EX(ai_CursorInterface_setTypeMap, 0, 0, 1)
+	ZEND_ARG_ARRAY_INFO(0, typemap, 0)
+ZEND_END_ARG_INFO()
+
+ZEND_BEGIN_ARG_INFO_EX(ai_CursorInterface_void, 0, 0, 0)
+ZEND_END_ARG_INFO()
+
+static zend_function_entry php_phongo_cursor_interface_me[] = {
+	/* clang-format off */
+	ZEND_ABSTRACT_ME(CursorInterface, getId, ai_CursorInterface_void)
+	ZEND_ABSTRACT_ME(CursorInterface, getServer, ai_CursorInterface_void)
+	ZEND_ABSTRACT_ME(CursorInterface, isDead, ai_CursorInterface_void)
```

Comment:
Before we codify this with an interface, what do you think about deprecating `isDead()` and introducing a new method in the next minor release (with this interface)? If so, this was justify a separate ticket to be done before this PR is finalized/merged.

This is based on [`mongoc_cursor_is_alive()`](http://mongoc.org/libmongoc/current/mongoc_cursor_is_alive.html), which libmongoc has superseded with [`mongoc_cursor_more()`](http://mongoc.org/libmongoc/current/mongoc_cursor_more.html). I think the documentation of `more()` is easier to understand for users:

> This function shall indicate if there is potentially more data to be read from the cursor. This is only useful with tailable cursors. Use mongoc_cursor_next() for regular cursors.

Since the underlying logic in libmongoc is based entirely on there _potentially_ being more results (based on the cursor's batch position and whether we have still have a cursor ID), a rename may better serve users who might otherwise assume that "alive" or "dead" are absolute.

Reply:
I've not done that yet though.

Reply:
Due to the nature of a cursor iteration, I don't think we can do better than _potentially_. The cases where a cursor _may_ have more data are as follows:

 * We know there are additional documents in the current batch (buffered on the client side). Iterating will fetch a document from our local buffer.
 * There are no additional documents in the current batch, but we have a cursor ID. Iterating will request the server for more documents, at which point we will either:
   *  Receive a batch containing more documents and the cursor ID will be kept open. This is common to iterating through a large result set.
   * Receive a batch containing more documents and the cursor ID will be zero, which signifies that we've received the final batch from the server (i.e. cursor is closed server-side)
   * Receive an empty batch from the server, but the cursor ID is kept open. This tells us we should consult the server for more documents later (common in tailable cursors).
   * Receive an empty batch from the server and the cursor ID will be zero. AFAIK, this is the only case where we are assured that there are not _potentially_ more documents (i.e. `more()` returns false).

Reply:
Those bullets were in response to your earlier comment:

> What we should have is a method that informs the user whether there is more data available in the cursor, and not just potentially.

I was trying to reason out why I don't think it's possible to provide such a method; although I don't know if you were advocating we introduce one or just wishing it were possible to do so. 

While brain-dumping the above points into [PHPC-1279](https://jira.mongodb.org/browse/PHPC-1279), I gave this some more thought and reconsidered `isDead()`. It seems like the "dead" case is simpler to understand (and probably document) than the "potentially more or alive" case. If you agree or don't feel strongly I'd be LGTM on leaving `isDead()` on the interface and simply improving its documentation down the line.